finality-aleph/src/sync/handler/request_handler.rs

```diff
@@ -0,0 +1,423 @@
+use core::{default::Default, marker::PhantomData};
+use std::{
+    collections::VecDeque,
+    fmt::{Display, Formatter},
+};
+
+use primitives::BlockNumber;
+
+use crate::{
+    session::{SessionBoundaryInfo, SessionId},
+    sync::{
+        data::BranchKnowledge, handler::Request, Block, BlockIdFor, BlockStatus, ChainStatus,
+        FinalizationStatus, Header, Justification,
+    },
+    BlockIdentifier,
+};
+
+#[derive(Debug, Clone)]
+pub enum RequestHandlerError<T: Display> {
+    RootMismatch,
+    MissingParent,
+    BadState,
+    BadRequest,
+    ChainStatusBadState,
+    ChainStatusError(T),
+}
+
+impl<T: Display> From<T> for RequestHandlerError<T> {
+    fn from(value: T) -> Self {
+        RequestHandlerError::ChainStatusError(value)
+    }
+}
+
+impl<T: Display> Display for RequestHandlerError<T> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            RequestHandlerError::RootMismatch => write!(f, "root mismatch"),
+            RequestHandlerError::MissingParent => write!(f, "missing parent"),
+            RequestHandlerError::ChainStatusError(e) => write!(f, "{}", e),
+            RequestHandlerError::BadState => write!(f, "bad state"),
+            RequestHandlerError::BadRequest => write!(f, "bad request"),
+            RequestHandlerError::ChainStatusBadState => write!(f, "bad state of chain_status"),
+        }
+    }
+}
+
+type HandlerResult<T, Error> = Result<T, RequestHandlerError<Error>>;
+type JustificationBlockAndHeader<J, B> =
+    (Option<J>, Option<B>, Option<<J as Justification>::Header>);
+type JustificationAndBlock<J, B> = (Option<J>, Option<B>);
+type StepResult<B, J> = (State, HeadOfChunk<B, J>, Vec<Chunk<B, J>>);
+
+pub enum Chunk<B, J>
+where
+    B: Block,
+    J: Justification<Header = B::Header>,
+{
+    Justification(J::Unverified),
+    Blocks(Vec<B>),
+    Headers(Vec<J::Header>),
+}
+
+pub struct Response<B, J>
+where
+    B: Block,
+    J: Justification<Header = B::Header>,
+{
+    pub chunks: Vec<Chunk<B, J>>,
+    pub block_to_request: Option<BlockIdFor<J>>,
```

Comment:
Eh, these seem to be exclusive, so this should be an `enum`, no?

Reply:
done